### DISCUSSION TITLE
Contract hyperparameters?

### DISCUSSION THREAD

**Discussion by NicolasRouquette** [pacti-org/pacti#235] [General]
In the space mission case study, I defined several functions to generate contracts using specific values of constant parameters, e.g:

```
# Parameters:
# - s: index of the timeline variables
# - generation: (min, max) rate of battery charge during the task instance
def CHARGING_power(s: int, generation: tuple[float, float]) -> PolyhedralContract:
  spec = PolyhedralContract.from_string(
    input_vars = [
      f"soc{s}_entry",          # initial battery SOC
      f"duration_charging{s}",  # variable task duration
    ],
    output_vars = [
      f"soc{s}_exit",           # final battery SOC
    ],
    assumptions = [
      # Task has a positive scheduled duration
      f"-duration_charging{s} <= 0",

      # Battery SOC must be positive
      f"-soc{s}_entry <= 0",
    ],
    guarantees = [
      # duration*generation(min) <= soc{exit} - soc{entry} <= duration*generation(max)
      f" soc{s}_exit - soc{s}_entry - {generation[1]}*duration_charging{s} <= 0",
      f"-soc{s}_exit + soc{s}_entry + {generation[0]}*duration_charging{s} <= 0",

      # Battery cannot exceed maximum SOC
      f"soc{s}_exit <= 100.0",
      
      # Battery should not completely discharge
      f"-soc{s}_exit <= 0",
    ])
  return spec
```

These function arguments define contract hyperparameters that would be useful for generating contracts using statistical sampling techniques to vary how many distinct instances of this contract to compose (i.e. the `s` argument) and the generation range for each instance.

**Comment by iincer** [comment]
This makes sense to me. What part of the functionality do you think should be part of Pacti vs. part of an external algorithm that calls Pacti?

**Reply by NicolasRouquette** [reply]
The essence involves distinguishing variables from hyper parameters in terms. It seems that this distinction ought to be in a layer that extends Pacti instead of being part of Pacti.
